rust/extractor/src/translate.rs

```diff
@@ -80,6 +156,671 @@ impl CrateTranslator<'_> {
+            id: TrapId::Star,
+            location,
+            pat,
+            guard,
+            expr,
+        })
+    }
+    fn emit_stmt(
+        &mut self,
+        stmt: &Statement,
+        body: &Body,
+        source_map: &BodySourceMap,
+    ) -> trap::Label {
+        match stmt {
+            Statement::Let {
+                pat,
+                type_ref,
+                initializer,
+                else_branch,
+            } => {
+                // TODO: find a way to get the location of the entire statement
+                let location = self.emit_location_pat(*pat, source_map);
+                let pat = self.emit_pat(*pat, body, source_map);
+                let type_ref = type_ref
+                    .as_ref()
+                    .map(|type_ref| self.emit_type_ref(type_ref));
+                let initializer =
+                    initializer.map(|initializer| self.emit_expr(initializer, body, source_map));
+                let else_ = else_branch.map(|else_| self.emit_expr(else_, body, source_map));
+
+                self.trap.emit(generated::IfLet {
+                    id: TrapId::Star,
+                    location,
+                    pat,
+                    type_ref,
+                    initializer,
+                    else_,
+                })
+            }
+            Statement::Expr { expr, has_semi } => {
+                let location = self.emit_location_expr(*expr, source_map);
+                let expr = self.emit_expr(*expr, body, source_map);
+                self.trap.emit(generated::ExprStmt {
+                    id: TrapId::Star,
+                    location,
+                    expr,
+                    has_semi: *has_semi,
+                })
+            }
+            Statement::Item => self.trap.emit(generated::ItemStmt {
+                id: TrapId::Star,
+                location: None,
+            }),
+        }
+    }
+    fn emit_expr(
+        &mut self,
+        expr_id: ExprId,
+        body: &Body,
+        source_map: &BodySourceMap,
+    ) -> trap::Label {
+        let location: Option<trap::Label> = self.emit_location_expr(expr_id, source_map);
+        let expr = &body[expr_id];
+        match expr {
+            ra_ap_hir_def::hir::Expr::Missing => self.trap.emit(generated::MissingExpr {
+                id: TrapId::Star,
+                location,
+            }),
+            ra_ap_hir_def::hir::Expr::Path(_) => self.trap.emit(generated::Path {
+                id: TrapId::Star,
+                location,
+            }),
+            ra_ap_hir_def::hir::Expr::If {
+                condition,
+                then_branch,
+                else_branch,
+            } => {
+                let condition = self.emit_expr(*condition, &body, &source_map);
+                let then = self.emit_expr(*then_branch, &body, &source_map);
+                let else_ = else_branch.map(|x| self.emit_expr(x, &body, &source_map));
+                self.trap.emit(generated::If {
```

Comment:
Maybe we want to suffix all `Expr` on the QL/codegen side? I.e. `IfExpr`, `LetExpr`, etc? You did that with `Pat`, migth be worth doing it for `Expr` and `Stmt` as well.

Reply:
Yes, that is probably most consistent. I had to suffix some of the others to avoid name clashes.